batch/Scouting/NtupleMaker/plugins/HitMaker.cc

```diff
@@ -88,10 +141,98 @@ void HitMaker::produce(edm::Event& iEvent, const edm::EventSetup& iSetup){
+                std::cout << " inside GL " << comp->surface().bounds().inside(refCenterGL)<< std::endl;
+                std::cout << " inside GLE " << comp->surface().bounds().inside(refCenterGL,myLocalErr)<< std::endl;
+            }
+         }
+    }
+    }
+    
+    for (auto const& vertex : *dvHandle) {
+        bool vertex_onmodule = false;
+        bool vertex_onmodule_withinunc = false;
+        GlobalPoint closestDetV(0,0,0);
+        float mindistance=10E6;
+        GlobalPoint myPoint(vertex.x(),vertex.y(),vertex.z());
+        GlobalError myErr(vertex.xError(),0,vertex.yError(),0,0,vertex.zError()); // cov missing
+        for (auto const& detL : searchGeom.allLayers()){
+            auto const& components = detL->basicComponents();
+            for (auto const& comp: components){
+                if(!comp->isLeaf()){
+                   for (auto const& subcomp: comp->components()){ 
+                       // not going further down beacause this level should be enough to separate all components (check above)
+                       Local3DPoint myLocalPoint = subcomp->surface().toLocal(myPoint);
+                       LocalError myLocalErr = ErrorFrameTransformer().transform( myErr, subcomp->surface());
+                       vertex_onmodule=subcomp->surface().bounds().inside(myLocalPoint);
+                       vertex_onmodule_withinunc=subcomp->surface().bounds().inside(myLocalPoint,myLocalErr);
+                       GlobalPoint retPoint(0,0,0);
+                       float mindist_temp=getMinDetDistance(subcomp, myLocalPoint, retPoint);
+                       if(mindist_temp<mindistance){mindistance=mindist_temp; closestDetV=retPoint;}
+                   }
+                }
+                else{
+                   Local3DPoint myLocalPoint = comp->surface().toLocal(myPoint);
+                   LocalError myLocalErr = ErrorFrameTransformer().transform( myErr, comp->surface());
+                   vertex_onmodule=comp->surface().bounds().inside(myLocalPoint);
+                   vertex_onmodule_withinunc=comp->surface().bounds().inside(myLocalPoint,myLocalErr);
+                   GlobalPoint retPoint(0,0,0);
+                   float mindist_temp=getMinDetDistance(comp, myLocalPoint, retPoint);
+                   if(mindist_temp<mindistance){mindistance=mindist_temp; closestDetV=retPoint;}
+                }
+            }
+            if(vertex_onmodule && vertex_onmodule_withinunc) break;     
+        }
+        dv_onmodule->push_back(vertex_onmodule);
+        dv_onmodule_withinunc->push_back(vertex_onmodule_withinunc);   
+        dv_detxmindr->push_back(closestDetV.x());
+        dv_detymindr->push_back(closestDetV.y());
+        dv_detzmindr->push_back(closestDetV.z());
+        
+    }
+   
     for (auto const& muon : *muonHandle) {
         vector<int> vertex_indices = muon.vtxIndx();
         int first_good_index = 0;
```

Comment:
This line should be followed by an intermediate loop over all `vertex_indices`:

```
int best_index = -1;
float maxprob = -1.0;
for (auto idx : vertex_indices) {
    if (idx >=0 && idx < (*dvHandle).size()) {
        Run3ScoutingVertex dv_ = (*dvHandle).at(idx);
        float prob_ = TMath::Prob(dv_.chi2(),dv_.ndof());
        if (prob_ > maxprob) {
            maxprob = prob_;
            best_index = idx;
        }
    }
}
```
Then, at LL.249-250, should replace `first_good_index` with `best_index`.
L. 249 would become:
```
if (best_index>=0) {
```
L. 250 would become:
```
Run3ScoutingVertex dv = (*dvHandle).at(best_index);
```

Reply:
I added that part. should be good.

Reply:
@leonardogiannini, if we keep `best_index` default to `0`, it may be that the script will crash in events with no DV (e.g., if we don't skim). So, I still think it's best to initialize it to `-1`, then just check for `if (best_index>=0)` later on - since it's already by definition < nDV?

Reply:
Done in https://github.com/cmstas/run3_scouting/pull/14/commits/0b53e31e6b2c7377f2ae84f974b9cb175f408e28